File: Python/34.Find_First_and_Last_Position_ofElement_inSortedArray.py
# 34. Find First and Last Position of Element in Sorted Array.

# Topic: Array, Binary Search.

'''
# Task:
-----------------
Given an array of integers nums sorted in non-decreasing order, find the starting and 
ending position of a given target value.

If target is not found in the array, return [-1, -1].

You must write an algorithm with O(log n) runtime complexity.

Example 1:
Input: nums = [5,7,7,8,8,10], target = 8
Output: [3,4]

Example 2:
Input: nums = [5,7,7,8,8,10], target = 6
Output: [-1,-1]

Example 3:
Input: nums = [], target = 0
Output: [-1,-1] 

Constraints:
0 <= nums.length <= 10^5
-10^9 <= nums[i] <= 10^9
nums is a non-decreasing array.
-10^9 <= target <= 10^9


# Testcase:
-----------------
[5,7,7,8,8,10]
8
[5,7,7,8,8,10]
6
[]
0


# Code:
-----------------
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
           
'''

# Solution:
from typing import List

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        # Define a binary search function to find the leftmost (first) occurrence of the target
        def binarySearchLeft(nums, target):
            left, right = 0, len(nums) - 1
            while left <= right:
                mid = (left + right) // 2
                # If the target is greater than the middle element, discard the left half
                if nums[mid] < target:
                    left = mid + 1
                # Otherwise, discard the right half
                else:
                    right = mid - 1
            # Return the left index where the target would be or the first occurrence
            return left

        # Define a binary search function to find the rightmost (last) occurrence of the target
        def binarySearchRight(nums, target):
            left, right = 0, len(nums) - 1
            while left <= right:
                mid = (left + right) // 2
                # If the target is greater or equal to the middle element, discard the left half
                if nums[mid] <= target:
                    left = mid + 1
                # Otherwise, discard the right half
                else:
                    right = mid - 1
            # Return the right index where the target's last occurrence is
            return right

        # Perform binary searches to find the first and last positions of the target
        left, right = binarySearchLeft(nums, target), binarySearchRight(nums, target)

        # Check if the target exists in the array
        if left <= right:
            # If it exists, return the range [left, right]
            return [left, right]
        # If the target is not found, return [-1, -1]
        return [-1, -1]

# Test cases
sol = Solution()
print(sol.searchRange([5,7,7,8,8,10], 8))  # Output: [3,4]
print(sol.searchRange([5,7,7,8,8,10], 6))  # Output: [-1,-1]
print(sol.searchRange([], 0))             # Output: [-1,-1]


# Description:
'''
To solve this problem, we can use a modified binary search approach to find the first and last 
positions of the target value in the sorted array. Since the array is sorted, binary search is 
suitable for achieving the required O(log n) runtime complexity.

The solution involves two main steps:

1. Finding the First Position: Perform binary search to find the first occurrence of the target. 
   In this binary search, when the target is found, continue searching in the left half to see if 
   there is an earlier occurrence of the target.

2. Finding the Last Position: Perform a second binary search to find the last occurrence of the target. 
   This time, when the target is found, continue searching in the right half to find the last occurrence.

In this code, binarySearchLeft and binarySearchRight are two binary search functions tailored to find the 
first and last occurrences of the target in the sorted array, respectively. After finding these positions, 
the code checks if the target actually exists in the array. If it does, the indices of the first and last 
occurrences are returned; otherwise, [-1, -1] is returned, indicating that the target is not present in 
the array. The solution thus effectively uses binary search to achieve the required O(log n) runtime 
complexity.
   
'''

File: Cpp/34. Find First and Last Position of Element in Sorted Array.cpp
// 34. Find First and Last Position of Element in Sorted Array.


// Topic: Array, Binary Search.


/*
### Task:
---
Given an array of integers nums sorted in non-decreasing order, find the starting and ending position of a given target value.

If target is not found in the array, return [-1, -1].

You must write an algorithm with O(log n) runtime complexity.

Example 1:
Input: nums = [5,7,7,8,8,10], target = 8
Output: [3,4]

Example 2:
Input: nums = [5,7,7,8,8,10], target = 6
Output: [-1,-1]

Example 3:
Input: nums = [], target = 0
Output: [-1,-1]

Constraints:
0 <= nums.length <= 10^5
-10^9 <= nums[i] <= 10^9
nums is a non-decreasing array.
-10^9 <= target <= 10^9


### Testcase:
---
[5,7,7,8,8,10]
8
[5,7,7,8,8,10]
6
[]
0


### Code:
---
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        
    }
};

*/
// Solution: --------------------------------------

#include <vector>
using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> result{-1, -1};  // Initialize the result with default values [-1, -1]

        // Perform binary search to find the first occurrence of the target
        int leftIndex = findFirstOrLast(nums, target, true);

        // If target is not found (leftIndex is out of bounds or does not point to target), return default result
        if (leftIndex == nums.size() || nums[leftIndex] != target) {
            return result;  // Target not found
        }

        // Perform binary search to find the last occurrence of the target
        // Subtracting 1 because the function returns the insertion point for the target, which is one index past the last occurrence
        int rightIndex = findFirstOrLast(nums, target, false) - 1;

        // Update result with the indices of the first and last occurrence
        result[0] = leftIndex;  // First position where target is found
        result[1] = rightIndex; // Last position where target is found

        return result;  // Return the updated result
    }

private:
    // Helper function to perform binary search
    // findFirst controls whether to find the first occurrence (true) or the insertion point for the last occurrence (false)
    int findFirstOrLast(vector<int>& nums, int target, bool findFirst) {
        int start = 0, end = nums.size();  // Initialize binary search bounds
        int mid;

        while (start < end) {  // Continue until search space is exhausted
            mid = start + (end - start) / 2;  // Prevent potential overflow

            // When finding the first occurrence, move end if target is found to continue searching left
            // When finding the last occurrence, only move end if target is strictly greater
            if (nums[mid] > target || (findFirst && nums[mid] == target)) {
                end = mid;  // Narrow search to the left half
            } else {
                start = mid + 1;  // Narrow search to the right half
            }
        }

        // For the first occurrence, start will point to the first target instance
        // For the last occurrence, start will point to the insertion point, which is one index past the last target instance
        return start;
    }
};



// Description: ===================================
/*
To solve the problem of finding the first and last position of an element in a sorted array with O(log n) runtime complexity, 
we can use a binary search algorithm. Since the array is sorted, binary search is an efficient way to find the target elements. 
However, instead of stopping at the first occurrence of the target, we need to find the first and last positions where this 
target appears.

The approach involves performing binary search twice with slight modifications:

1. First binary search to find the first (leftmost) occurrence of the target.
2. Second binary search to find the last (rightmost) occurrence of the target.


### Detailed Explanation:

- The function `searchRange` initializes the result vector with `[-1, -1]`, which will be returned if the target is not found.
- `findFirstOrLast` is a helper function that performs a binary search to find the first or last occurrence of the target. It takes 
   a boolean parameter `findFirst` to indicate whether to find the first or last occurrence.

- In the first call to `findFirstOrLast`, it searches for the first occurrence by setting `findFirst` to `true`. If the target is 
  not found (i.e., `leftIndex` is equal to the size of the array or the value at `leftIndex` is not the target), the function 
  returns `[-1, -1]`.

- If the first occurrence is found, the function calls `findFirstOrLast` again with `findFirst` set to `false` to find the last 
  occurrence. The last occurrence's index is one less than the returned value because `findFirstOrLast` returns the index where 
  the target could be inserted to maintain the order, which is one position after the last occurrence of the target.

- The result vector is then updated with the first and last positions of the target and returned.

*/
